// @pnp/odata/dist/odata.es5.umd.js
/**
 * @license
 * v1.3.7
 * MIT (https://github.com/pnp/pnpjs/blob/master/LICENSE)
 * Copyright (c) 2019 Microsoft
 * docs: https://pnp.github.io/pnpjs/
 * source: https://github.com/pnp/pnpjs
 * bugs: https://github.com/pnp/pnpjs/issues
 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@pnp/common'), require('@pnp/logging')) :
    typeof define === 'function' && define.amd ? define(['exports', '@pnp/common', '@pnp/logging'], factory) :
    (factory((global.pnp = global.pnp || {}, global.pnp.odata = {}),global.pnp.common,global.pnp.logging));
}(this, (function (exports,common,logging) { 'use strict';

    var CachingOptions = /** @class */ (function () {
        function CachingOptions(key) {
            this.key = key;
            this.expiration = common.dateAdd(new Date(), "second", common.RuntimeConfig.defaultCachingTimeoutSeconds);
            this.storeName = common.RuntimeConfig.defaultCachingStore;
        }
        Object.defineProperty(CachingOptions.prototype, "store", {
            get: function () {
                if (this.storeName === "local") {
                    return CachingOptions.storage.local;
                }
                else {
                    return CachingOptions.storage.session;
                }
            },
            enumerable: true,
            configurable: true
        });
        CachingOptions.storage = new common.PnPClientStorage();
        return CachingOptions;
    }());
    var CachingParserWrapper = /** @class */ (function () {
        function CachingParserWrapper(parser, cacheOptions) {
            this.parser = parser;
            this.cacheOptions = cacheOptions;
        }
        CachingParserWrapper.prototype.parse = function (response) {
            var _this = this;
            return this.parser.parse(response).then(function (r) { return _this.cacheData(r); });
        };
        CachingParserWrapper.prototype.cacheData = function (data) {
            if (this.cacheOptions.store !== null) {
                this.cacheOptions.store.put(this.cacheOptions.key, data, this.cacheOptions.expiration);
            }
            return data;
        };
        return CachingParserWrapper;
    }());

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    var HttpRequestError = /** @class */ (function (_super) {
        __extends(HttpRequestError, _super);
        function HttpRequestError(message, response, status, statusText) {
            if (status === void 0) { status = response.status; }
            if (statusText === void 0) { statusText = response.statusText; }
            var _this = _super.call(this, message) || this;
            _this.response = response;
            _this.status = status;
            _this.statusText = statusText;
            _this.isHttpRequestError = true;
            return _this;
        }
        HttpRequestError.init = function (r) {
            return r.clone().text().then(function (t) {
                return new HttpRequestError("Error making HttpClient request in queryable [" + r.status + "] " + r.statusText + " ::> " + t, r.clone());
            });
        };
        return HttpRequestError;
    }(Error));
    var ODataParserBase = /** @class */ (function () {
        function ODataParserBase() {
            this.rawJson = {};
        }
        ODataParserBase.prototype.parse = function (r) {
            var _this = this;
            return new Promise(function (resolve, reject) {
                if (_this.handleError(r, reject)) {
                    _this.parseImpl(r, resolve, reject);
                }
            });
        };
        ODataParserBase.prototype.parseImpl = function (r, resolve, reject) {
            var _this = this;
            if ((r.headers.has("Content-Length") && parseFloat(r.headers.get("Content-Length")) === 0) || r.status === 204) {
                resolve({});
            }
            else {
                // patch to handle cases of 200 response with no or whitespace only bodies (#487 & #545)
                r.text()
                    .then(function (txt) { return txt.replace(/\s/ig, "").length > 0 ? JSON.parse(txt) : {}; })
                    .then(function (json) { return resolve(_this.parseODataJSON(json)); })
                    .catch(function (e) { return reject(e); });
            }
        };
        /**
         * Handles a response with ok === false by parsing the body and creating a ProcessHttpClientResponseException
         * which is passed to the reject delegate. This method returns true if there is no error, otherwise false
         *
         * @param r Current response object
         * @param reject reject delegate for the surrounding promise
         */
        ODataParserBase.prototype.handleError = function (r, reject) {
            if (!r.ok) {
                HttpRequestError.init(r).then(reject);
            }
            return r.ok;
        };
        /**
         * Normalizes the json response by removing the various nested levels
         *
         * @param json json object to parse
         */
        ODataParserBase.prototype.parseODataJSON = function (json) {
            this.rawJson = json;
            var result = json;
            if (common.hOP(json, "d")) {
                if (common.hOP(json.d, "results")) {
                    result = json.d.results;
                }
                else {
                    result = json.d;
                }
            }
            else if (common.hOP(json, "value")) {
                result = json.value;
            }
            return result;
        };
        return ODataParserBase;
    }());
    var ODataDefaultParser = /** @class */ (function (_super) {
        __extends(ODataDefaultParser, _super);
        function ODataDefaultParser() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        return ODataDefaultParser;
    }(ODataParserBase));
    var TextParser = /** @class */ (function (_super) {
        __extends(TextParser, _super);
        function TextParser() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        TextParser.prototype.parseImpl = function (r, resolve) {
            r.text().then(resolve);
        };
        return TextParser;
    }(ODataParserBase));
    var BlobParser = /** @class */ (function (_super) {
        __extends(BlobParser, _super);
        function BlobParser() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        BlobParser.prototype.parseImpl = function (r, resolve) {
            r.blob().then(resolve);
        };
        return BlobParser;
    }(ODataParserBase));
    var JSONParser = /** @class */ (function (_super) {
        __extends(JSONParser, _super);
        function JSONParser() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        JSONParser.prototype.parseImpl = function (r, resolve) {
            r.json().then(resolve);
        };
        return JSONParser;
    }(ODataParserBase));
    var BufferParser = /** @class */ (function (_super) {
        __extends(BufferParser, _super);
        function BufferParser() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        BufferParser.prototype.parseImpl = function (r, resolve) {
            if (common.isFunc(r.arrayBuffer)) {
                r.arrayBuffer().then(resolve);
            }
            else {
                r.buffer().then(resolve);
            }
        };
        return BufferParser;
    }(ODataParserBase));
    var LambdaParser = /** @class */ (function (_super) {
        __extends(LambdaParser, _super);
        function LambdaParser(parser) {
            var _this = _super.call(this) || this;
            _this.parser = parser;
            return _this;
        }
        LambdaParser.prototype.parseImpl = function (r, resolve) {
            this.parser(r).then(resolve);
        };
        return LambdaParser;
    }(ODataParserBase));

    /**
     * Resolves the context's result value
     *
     * @param context The current context
     */
    function returnResult(context) {
        logging.Logger.log({
            data: logging.Logger.activeLogLevel === 0 /* Verbose */ ? context.result : {},
            level: 1 /* Info */,
            message: "[" + context.requestId + "] (" + (new Date()).getTime() + ") Returning result from pipeline. Set logging to verbose to see data.",
        });
        return Promise.resolve(context.result);
    }
    /**
     * Sets the result on the context
     */
    function setResult(context, value) {
        return new Promise(function (resolve) {
            context.result = value;
            context.hasResult = true;
            resolve(context);
        });
    }
    /**
     * Invokes the next method in the provided context's pipeline
     *
     * @param c The current request context
     */
    function next(c) {
        if (c.pipeline.length > 0) {
            return c.pipeline.shift()(c);
        }
        else {
            return Promise.resolve(c);
        }
    }
    /**
     * Executes the current request context's pipeline
     *
     * @param context Current context
     */
    function pipe(context) {
        if (context.pipeline.length < 1) {
            logging.Logger.write("[" + context.requestId + "] (" + (new Date()).getTime() + ") Request pipeline contains no methods!", 2 /* Warning */);
        }
        var promise = next(context).then(function (ctx) { return returnResult(ctx); }).catch(function (e) {
            logging.Logger.error(e);
            throw e;
        });
        if (context.isBatched) {
            // this will block the batch's execute method from returning until the child requets have been resolved
            context.batch.addResolveBatchDependency(promise);
        }
        return promise;
    }
    /**
     * decorator factory applied to methods in the pipeline to control behavior
     */
    function requestPipelineMethod(alwaysRun) {
        if (alwaysRun === void 0) { alwaysRun = false; }
        return function (target, propertyKey, descriptor) {
            var method = descriptor.value;
            descriptor.value = function () {
                var args = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    args[_i] = arguments[_i];
                }
                // if we have a result already in the pipeline, pass it along and don't call the tagged method
                if (!alwaysRun && args.length > 0 && common.hOP(args[0], "hasResult") && args[0].hasResult) {
                    logging.Logger.write("[" + args[0].requestId + "] (" + (new Date()).getTime() + ") Skipping request pipeline method " + propertyKey + ", existing result in pipeline.", 0 /* Verbose */);
                    return Promise.resolve(args[0]);
                }
                // apply the tagged method
                logging.Logger.write("[" + args[0].requestId + "] (" + (new Date()).getTime() + ") Calling request pipeline method " + propertyKey + ".", 0 /* Verbose */);
                // then chain the next method in the context's pipeline - allows for dynamic pipeline
                return method.apply(target, args).then(function (ctx) { return next(ctx); });
            };
        };
    }
    /**
     * Contains the methods used within the request pipeline
     */
    var PipelineMethods = /** @class */ (function () {
        function PipelineMethods() {
        }
        /**
         * Logs the start of the request
         */
        PipelineMethods.logStart = function (context) {
            return new Promise(function (resolve) {
                logging.Logger.log({
                    data: logging.Logger.activeLogLevel === 1 /* Info */ ? {} : context,
                    level: 1 /* Info */,
                    message: "[" + context.requestId + "] (" + (new Date()).getTime() + ") Beginning " + context.verb + " request (" + context.requestAbsoluteUrl + ")",
                });
                resolve(context);
            });
        };
        /**
         * Handles caching of the request
         */
        PipelineMethods.caching = function (context) {
            return new Promise(function (resolve) {
                // handle caching, if applicable
                if (context.isCached) {
                    logging.Logger.write("[" + context.requestId + "] (" + (new Date()).getTime() + ") Caching is enabled for request, checking cache...", 1 /* Info */);
                    var cacheOptions = new CachingOptions(context.requestAbsoluteUrl.toLowerCase());
                    if (context.cachingOptions !== undefined) {
                        cacheOptions = common.extend(cacheOptions, context.cachingOptions);
                    }
                    // we may not have a valid store
                    if (cacheOptions.store !== null) {
                        // check if we have the data in cache and if so resolve the promise and return
                        var data = cacheOptions.store.get(cacheOptions.key);
                        if (data !== null) {
                            // ensure we clear any held batch dependency we are resolving from the cache
                            logging.Logger.log({
                                data: logging.Logger.activeLogLevel === 1 /* Info */ ? {} : data,
                                level: 1 /* Info */,
                                message: "[" + context.requestId + "] (" + (new Date()).getTime() + ") Value returned from cache.",
                            });
                            if (common.isFunc(context.batchDependency)) {
                                context.batchDependency();
                            }
                            // handle the case where a parser needs to take special actions with a cached result
                            if (common.hOP(context.parser, "hydrate")) {
                                data = context.parser.hydrate(data);
                            }
                            return setResult(context, data).then(function (ctx) { return resolve(ctx); });
                        }
                    }
                    logging.Logger.write("[" + context.requestId + "] (" + (new Date()).getTime() + ") Value not found in cache.", 1 /* Info */);
                    // if we don't then wrap the supplied parser in the caching parser wrapper
                    // and send things on their way
                    context.parser = new CachingParserWrapper(context.parser, cacheOptions);
                }
                return resolve(context);
            });
        };
        /**
         * Sends the request
         */
        PipelineMethods.send = function (context) {
            return new Promise(function (resolve, reject) {
                // send or batch the request
                if (context.isBatched) {
                    // we are in a batch, so add to batch, remove dependency, and resolve with the batch's promise
                    var p = context.batch.add(context.requestAbsoluteUrl, context.verb, context.options, context.parser, context.requestId);
                    // we release the dependency here to ensure the batch does not execute until the request is added to the batch
                    if (common.isFunc(context.batchDependency)) {
                        context.batchDependency();
                    }
                    logging.Logger.write("[" + context.requestId + "] (" + (new Date()).getTime() + ") Batching request in batch " + context.batch.batchId + ".", 1 /* Info */);
                    // we set the result as the promise which will be resolved by the batch's execution
                    resolve(setResult(context, p));
                }
                else {
                    logging.Logger.write("[" + context.requestId + "] (" + (new Date()).getTime() + ") Sending request.", 1 /* Info */);
                    // we are not part of a batch, so proceed as normal
                    var client = context.clientFactory();
                    var opts = common.extend(context.options || {}, { method: context.verb });
                    client.fetch(context.requestAbsoluteUrl, opts)
                        .then(function (response) { return context.parser.parse(response); })
                        .then(function (result) { return setResult(context, result); })
                        .then(function (ctx) { return resolve(ctx); })
                        .catch(function (e) { return reject(e); });
                }
            });
        };
        /**
         * Logs the end of the request
         */
        PipelineMethods.logEnd = function (context) {
            return new Promise(function (resolve) {
                if (context.isBatched) {
                    logging.Logger.log({
                        data: logging.Logger.activeLogLevel === 1 /* Info */ ? {} : context,
                        level: 1 /* Info */,
                        message: "[" + context.requestId + "] (" + (new Date()).getTime() + ") " + context.verb + " request will complete in batch " + context.batch.batchId + ".",
                    });
                }
                else {
                    logging.Logger.log({
                        data: logging.Logger.activeLogLevel === 1 /* Info */ ? {} : context,
                        level: 1 /* Info */,
                        message: "[" + context.requestId + "] (" + (new Date()).getTime() + ") Completing " + context.verb + " request.",
                    });
                }
                resolve(context);
            });
        };
        __decorate([
            requestPipelineMethod(true)
        ], PipelineMethods, "logStart", null);
        __decorate([
            requestPipelineMethod()
        ], PipelineMethods, "caching", null);
        __decorate([
            requestPipelineMethod()
        ], PipelineMethods, "send", null);
        __decorate([
            requestPipelineMethod(true)
        ], PipelineMethods, "logEnd", null);
        return PipelineMethods;
    }());
    function getDefaultPipeline() {
        return [
            PipelineMethods.logStart,
            PipelineMethods.caching,
            PipelineMethods.send,
            PipelineMethods.logEnd,
        ].slice(0);
    }

    var Queryable = /** @class */ (function () {
        function Queryable() {
            this._query = new Map();
            this._options = {};
            this._url = "";
            this._parentUrl = "";
            this._useCaching = false;
            this._cachingOptions = null;
            this._cloneParentWasCaching = false;
            this._cloneParentCacheOptions = null;
            this._requestPipeline = null;
        }
        /**
        * Gets the currentl url
        *
        */
        Queryable.prototype.toUrl = function () {
            return this._url;
        };
        /**
         * Directly concatonates the supplied string to the current url, not normalizing "/" chars
         *
         * @param pathPart The string to concatonate to the url
         */
        Queryable.prototype.concat = function (pathPart) {
            this._url += pathPart;
            return this;
        };
        Object.defineProperty(Queryable.prototype, "query", {
            /**
             * Provides access to the query builder for this url
             *
             */
            get: function () {
                return this._query;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * Sets custom options for current object and all derived objects accessible via chaining
         *
         * @param options custom options
         */
        Queryable.prototype.configure = function (options) {
            common.mergeOptions(this._options, options);
            return this;
        };
        /**
         * Configures this instance from the configure options of the supplied instance
         *
         * @param o Instance from which options should be taken
         */
        Queryable.prototype.configureFrom = function (o) {
            common.mergeOptions(this._options, o._options);
            return this;
        };
        /**
         * Enables caching for this request
         *
         * @param options Defines the options used when caching this request
         */
        Queryable.prototype.usingCaching = function (options) {
            if (!common.RuntimeConfig.globalCacheDisable) {
                this._useCaching = true;
                if (options !== undefined) {
                    this._cachingOptions = options;
                }
            }
            return this;
        };
        /**
         * Allows you to set a request specific processing pipeline
         *
         * @param pipeline The set of methods, in order, to execute a given request
         */
        Queryable.prototype.withPipeline = function (pipeline) {
            this._requestPipeline = pipeline.slice(0);
            return this;
        };
        Queryable.prototype.getCore = function (parser, options) {
            if (parser === void 0) { parser = new JSONParser(); }
            if (options === void 0) { options = {}; }
            // Fix for #304 - when we clone objects we in some cases then execute a get request
            // in these cases the caching settings were getting dropped from the request
            // this tracks if the object from which this was cloned was caching and applies that to an immediate get request
            // does not affect objects cloned from this as we are using different fields to track the settings so it won't
            // be triggered
            if (this._cloneParentWasCaching) {
                this.usingCaching(this._cloneParentCacheOptions);
            }
            return this.reqImpl("GET", options, parser);
        };
        Queryable.prototype.postCore = function (options, parser) {
            if (options === void 0) { options = {}; }
            if (parser === void 0) { parser = new JSONParser(); }
            return this.reqImpl("POST", options, parser);
        };
        Queryable.prototype.patchCore = function (options, parser) {
            if (options === void 0) { options = {}; }
            if (parser === void 0) { parser = new JSONParser(); }
            return this.reqImpl("PATCH", options, parser);
        };
        Queryable.prototype.deleteCore = function (options, parser) {
            if (options === void 0) { options = {}; }
            if (parser === void 0) { parser = new JSONParser(); }
            return this.reqImpl("DELETE", options, parser);
        };
        Queryable.prototype.putCore = function (options, parser) {
            if (options === void 0) { options = {}; }
            if (parser === void 0) { parser = new JSONParser(); }
            return this.reqImpl("PUT", options, parser);
        };
        Queryable.prototype.reqImpl = function (method, options, parser) {
            var _this = this;
            if (options === void 0) { options = {}; }
            return this.getRequestPipeline(method, options, parser)
                .then(function (pipeline) { return _this.toRequestContext(method, options, parser, pipeline); })
                .then(function (context) { return pipe(context); });
        };
        /**
         * Appends the given string and normalizes "/" chars
         *
         * @param pathPart The string to append
         */
        Queryable.prototype.append = function (pathPart) {
            this._url = common.combine(this._url, pathPart);
        };
        Object.defineProperty(Queryable.prototype, "parentUrl", {
            /**
             * Gets the parent url used when creating this instance
             *
             */
            get: function () {
                return this._parentUrl;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * Extends this queryable from the provided parent
         *
         * @param parent Parent queryable from which we will derive a base url
         * @param path Additional path
         */
        Queryable.prototype.extend = function (parent, path) {
            this._parentUrl = parent._url;
            this._url = common.combine(this._parentUrl, path || "");
            this.configureFrom(parent);
        };
        /**
         * Configures a cloned object from this instance
         *
         * @param clone
         */
        Queryable.prototype._clone = function (clone, _0) {
            clone.configureFrom(this);
            if (this._useCaching) {
                clone._cloneParentWasCaching = true;
                clone._cloneParentCacheOptions = this._cachingOptions;
            }
            return clone;
        };
        /**
         * Handles getting the request pipeline to run for a given request
         */
        // @ts-ignore
        // justified because we want to show that all these arguments are passed to the method so folks inheriting and potentially overriding
        // clearly see how the method is invoked inside the class
        Queryable.prototype.getRequestPipeline = function (method, options, parser) {
            var _this = this;
            if (options === void 0) { options = {}; }
            return new Promise(function (resolve) {
                if (common.objectDefinedNotNull(_this._requestPipeline) && common.isArray(_this._requestPipeline)) {
                    resolve(_this._requestPipeline);
                }
                else {
                    resolve(getDefaultPipeline());
                }
            });
        };
        return Queryable;
    }());
    var ODataQueryable = /** @class */ (function (_super) {
        __extends(ODataQueryable, _super);
        function ODataQueryable() {
            var _this = _super.call(this) || this;
            _this._batch = null;
            _this._batchDependency = null;
            return _this;
        }
        /**
         * Adds this query to the supplied batch
         *
         * @example
         * ```
         *
         * let b = pnp.sp.createBatch();
         * pnp.sp.web.inBatch(b).get().then(...);
         * b.execute().then(...)
         * ```
         */
        ODataQueryable.prototype.inBatch = function (batch) {
            if (this.batch !== null) {
                throw Error("This query is already part of a batch.");
            }
            if (common.objectDefinedNotNull(batch)) {
                this._batch = batch;
            }
            return this;
        };
        /**
         * Gets the currentl url
         *
         */
        ODataQueryable.prototype.toUrl = function () {
            return this._url;
        };
        /**
         * Executes the currently built request
         *
         * @param parser Allows you to specify a parser to handle the result
         * @param getOptions The options used for this request
         */
        ODataQueryable.prototype.get = function (parser, options) {
            if (parser === void 0) { parser = new ODataDefaultParser(); }
            if (options === void 0) { options = {}; }
            return this.getCore(parser, options);
        };
        ODataQueryable.prototype.getCore = function (parser, options) {
            if (parser === void 0) { parser = new ODataDefaultParser(); }
            if (options === void 0) { options = {}; }
            return _super.prototype.getCore.call(this, parser, options);
        };
        ODataQueryable.prototype.postCore = function (options, parser) {
            if (options === void 0) { options = {}; }
            if (parser === void 0) { parser = new ODataDefaultParser(); }
            return _super.prototype.postCore.call(this, options, parser);
        };
        ODataQueryable.prototype.patchCore = function (options, parser) {
            if (options === void 0) { options = {}; }
            if (parser === void 0) { parser = new ODataDefaultParser(); }
            return _super.prototype.patchCore.call(this, options, parser);
        };
        ODataQueryable.prototype.deleteCore = function (options, parser) {
            if (options === void 0) { options = {}; }
            if (parser === void 0) { parser = new ODataDefaultParser(); }
            return _super.prototype.deleteCore.call(this, options, parser);
        };
        ODataQueryable.prototype.putCore = function (options, parser) {
            if (options === void 0) { options = {}; }
            if (parser === void 0) { parser = new ODataDefaultParser(); }
            return _super.prototype.putCore.call(this, options, parser);
        };
        ODataQueryable.prototype.reqImpl = function (method, options, parser) {
            if (options === void 0) { options = {}; }
            if (this.hasBatch) {
                this._batchDependency = this.addBatchDependency();
            }
            return _super.prototype.reqImpl.call(this, method, options, parser);
        };
        /**
         * Blocks a batch call from occuring, MUST be cleared by calling the returned function
        */
        ODataQueryable.prototype.addBatchDependency = function () {
            if (this._batch !== null) {
                return this._batch.addDependency();
            }
            return function () { return null; };
        };
        Object.defineProperty(ODataQueryable.prototype, "hasBatch", {
            /**
             * Indicates if the current query has a batch associated
             *
             */
            get: function () {
                return common.objectDefinedNotNull(this._batch);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ODataQueryable.prototype, "batch", {
            /**
             * The batch currently associated with this query or null
             *
             */
            get: function () {
                return this.hasBatch ? this._batch : null;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * Configures a cloned object from this instance
         *
         * @param clone
         */
        ODataQueryable.prototype._clone = function (clone, cloneSettings) {
            clone = _super.prototype._clone.call(this, clone, cloneSettings);
            if (cloneSettings.includeBatch) {
                clone = clone.inBatch(this._batch);
            }
            return clone;
        };
        return ODataQueryable;
    }(Queryable));

    var ODataBatch = /** @class */ (function () {
        function ODataBatch(_batchId) {
            if (_batchId === void 0) { _batchId = common.getGUID(); }
            this._batchId = _batchId;
            this._reqs = [];
            this._deps = [];
            this._rDeps = [];
        }
        Object.defineProperty(ODataBatch.prototype, "batchId", {
            get: function () {
                return this._batchId;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ODataBatch.prototype, "requests", {
            /**
             * The requests contained in this batch
             */
            get: function () {
                return this._reqs;
            },
            enumerable: true,
            configurable: true
        });
        /**
         *
         * @param url Request url
         * @param method Request method (GET, POST, etc)
         * @param options Any request options
         * @param parser The parser used to handle the eventual return from the query
         * @param id An identifier used to track a request within a batch
         */
        ODataBatch.prototype.add = function (url, method, options, parser, id) {
            var info = {
                id: id,
                method: method.toUpperCase(),
                options: options,
                parser: parser,
                reject: null,
                resolve: null,
                url: url,
            };
            var p = new Promise(function (resolve, reject) {
                info.resolve = resolve;
                info.reject = reject;
            });
            this._reqs.push(info);
            return p;
        };
        /**
         * Adds a dependency insuring that some set of actions will occur before a batch is processed.
         * MUST be cleared using the returned resolve delegate to allow batches to run
         */
        ODataBatch.prototype.addDependency = function () {
            var resolver = function () { return void (0); };
            this._deps.push(new Promise(function (resolve) {
                resolver = resolve;
            }));
            return resolver;
        };
        /**
         * The batch's execute method will not resolve util any promises added here resolve
         *
         * @param p The dependent promise
         */
        ODataBatch.prototype.addResolveBatchDependency = function (p) {
            this._rDeps.push(p);
        };
        /**
         * Execute the current batch and resolve the associated promises
         *
         * @returns A promise which will be resolved once all of the batch's child promises have resolved
         */
        ODataBatch.prototype.execute = function () {
            var _this = this;
            // we need to check the dependencies twice due to how different engines handle things.
            // We can get a second set of promises added during the first set resolving
            return Promise.all(this._deps)
                .then(function () { return Promise.all(_this._deps); })
                .then(function () { return _this.executeImpl(); })
                .then(function () { return Promise.all(_this._rDeps); })
                .then(function () { return void (0); });
        };
        return ODataBatch;
    }());

    exports.CachingOptions = CachingOptions;
    exports.CachingParserWrapper = CachingParserWrapper;
    exports.HttpRequestError = HttpRequestError;
    exports.ODataParserBase = ODataParserBase;
    exports.ODataDefaultParser = ODataDefaultParser;
    exports.TextParser = TextParser;
    exports.BlobParser = BlobParser;
    exports.JSONParser = JSONParser;
    exports.BufferParser = BufferParser;
    exports.LambdaParser = LambdaParser;
    exports.setResult = setResult;
    exports.pipe = pipe;
    exports.requestPipelineMethod = requestPipelineMethod;
    exports.PipelineMethods = PipelineMethods;
    exports.getDefaultPipeline = getDefaultPipeline;
    exports.Queryable = Queryable;
    exports.ODataQueryable = ODataQueryable;
    exports.ODataBatch = ODataBatch;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
